### ISSUE TITLE
Write plugin to allow custom uploads to GitHub releases

### ISSUE THREAD

**Issue by gr2m** [semantic-release/semantic-release#72]
For JavaScript libraries that can be used in the browser, it would be nice to have direct downloads of the different versions, ideally right from the GitHub releases page. Here is an example:
https://github.com/gr2m/initials/releases

It would be nice to be able to not only download the source code for each version, but also the built `initials.js` and maybe a`initials.min.js` files.

Could that be with a plugin?

**Comment by ariporad** [comment]
Hmm...




I'm working on a PR to seperate out the publishing (npm) and releasing (github) into plugins. One that lands, this will be much easier.




Why don't we re-discuss this once that lands?



Thanks!Ari

**Comment by boennemann** [comment]
@gr2m It should be easy to add the npm tarball, or any arbitrary file really, because we just need to pass it to the github.repos.createRelease method.

I'll think about how to make that configurable.

**Comment by gr2m** [comment]
I'm happy to give this a try, whenever the time is good for it, I'd just need a little guidance :)

**Comment by rodneyrehm** [comment]
+1

**Comment by gr2m** [comment]
@boennemann I’m still happy to work on this, but could use a little guidance on how to implement this. I didn’t touch semantic-release code yet. I don’t know what you mean with `pass it to the github.repos.createRelease`

**Comment by hbetts** [comment]
@gr2m I think @boennemann was referring to updating the `release` object here: https://github.com/semantic-release/semantic-release/blob/caribou/src/post.js#L28-L36

That object is passed to `createRelease` on line 47.

However, I think you actually want to add a new call to `uploadAsset`. `uploadAsset` is documented in the `node-github` API documentation here: http://mikedeboer.github.io/node-github/#releases.prototype.uploadAsset

@ariporad and @gr2m as part of #150 I pulled out the GitHub code into a separate plugin (though still bundled with `semantic-release`.

My hope is to support additional plugins like [semantic-release-github-notifier](https://github.com/hbetts/semantic-release-github-notifier).

I actually just started work on pulling out the GitHub code from #150 into a separate project [here](https://github.com/hbetts/semantic-release-github-releaser).

**Comment by gr2m** [comment]
@hbetts that sounds great. So I guess best would be to wait for your work to land in semantic-release, and then build a plugin based on it to allow custom uploads?

**Comment by finom** [comment]
Any news about this issue?

**Comment by hbetts** [comment]
I closed out my pull requests to integrate my changes into `semantic-release` since there was no feedback on them.

Instead, to support my own needs I moved forward with developing my own semantic-release like tool for GitLab.

**Comment by finom** [comment]
@hbetts this was awesome idea. Unfortunately semantic-release looks dead :(

**Comment by gr2m** [comment]
I use it in about 50 of my open source projects, it’s most definitely not dead, it works great for what it does, but it’s only as active as you are willing to contribute ;) Stephan who started the repo is pretty swamped with life right now, but if you take this and start a PR that will make it work, I’m sure we can get it added

**Comment by finom** [comment]
@gr2m I didn't say it is dead I said it looks dead. This is the great project as MVP but it needs to be rewritten from scratch or at least make big refactoring, dependencies update, issues review. I guess you like it because you use it as described at tutorials but I'm trying to customize it for my needs (for example I'm that guy who use plugins). I'm going to use it in some open source projects too but only because I have no alternatives.

**Comment by gr2m** [comment]
Yeah I agree it needs a bigger effort to refactor as it grew organically over time. I think it proofed it scope pretty good now. I already told Stephan I’m willing to help maintain semantic-release and help with such an update, are you in?

**Comment by finom** [comment]
@gr2m I can't say I'm in because mostly I'm busy with my own projects but I'm going to look on how is going and maybe provide something useful for the project.

**Comment by hbetts** [comment]
@gr2m the project, it's maintenance and community contributions, will only be as active as it's maintainer(s). For example, I would have liked to be more active in supporting `semantic-release`, but given that my merge requests (including #150 which was to help set the foundation for resolving this issue) were not reviewed, let alone accepted. That makes it difficult to be _active_.

There's actually an on-going discussion to address the plugin system, which you can read here: https://gist.github.com/boennemann/ac06b7f8bcf50510ffa1

Also, there's a [project-management](https://github.com/semantic-release/project-management) repo for discussing major refactors, ideas, re-tooling, etc.

I would have also liked to help with `semantic-release`, including providing code reviews, and project management support, but now I'm a little busy with my own `semantic-release` like project that works with GitLab.

**Comment by finom** [comment]
To keep a bundle up to date somewhere I have created [a little script](https://github.com/finom/deploy-to-git) that deploys it to given branch. It's more useful for me than custom uploads because I can use bundled files directly with gh-pages. This also allows to modify ``generateNotes`` script to mention that branch in release description. I'm not sure does README describe the tool nicely, let me know if you have suggestions.

P. S. I have spent really long hours on a research of how to deploy a bundle of a library and I found deploying to a branch of the same repo as the most useful and universal (considering that I don't like branches with different commit history).

**Comment by conblem** [comment]
Any News on this?

**Comment by gr2m** [comment]
I’m working on something by chance, I might have something in a week or two, I’ll post updates in the comments here

**Comment by gr2m** [comment]
oops forgot to post the library I’ve built: https://github.com/gr2m/octokit-release-asset-upload

but even more awesome: uploading assets will be a built-in feature soon! Thanks t @pvdlg 🙌

**Comment by conblem** [comment]
@gr2m Cool stuff! would love to see a example in conjunction with semantic-release :blush:

**Comment by gr2m** [comment]
@conblem see https://github.com/gr2m/octokit-release-asset-upload/blob/master/package.json#L21 (actually have to fix this to `postpublish` instead of `postsemantic-relaese` :) )

But note that the new semantic-release will have asset upload options built in :)

**Comment by pvdlg** [comment]
@conblem, We just release `semantic-release@10.0.0` on the dist-tag `next` that include the possibility to publish file to a Github release
You can find more information on that in the new github plugin: https://github.com/semantic-release/github#assets-option

Let us know if that works for you!
